### ISSUE TITLE
Blur/Frosted Glass support

### ISSUE THREAD

**Issue by WatchMkr** [pop-os/cosmic-comp#511]

**Comment by uiYzzi** [comment]
I need it

**Comment by Sythelux** [comment]
Adding "Aero", "Glossy" here for the search engines.

**Comment by anarchy1923** [comment]
That would be awesome to have!

**Comment by systemd11** [comment]
what happened?

**Comment by Sythelux** [comment]
> what happened?

it just means, that it was pushed back to the next release and won't be in the current one, which is a normal thing to do in agile software development

**Comment by Necryl** [comment]
Oh so this will not be implemented until the first full release?

**Comment by leviport** [comment]
Correct

**Comment by candsastle** [comment]
I would really like an option to adjust the blur strength.

**Comment by anarchy1923** [comment]
> I would really like an option to adjust the blur strength.

I guess gotta wait for Epoch 2 until it is implemented.

**Comment by Milli189** [comment]
I'm really looking forward for this to be added

**Comment by HenryNordin** [comment]
Also looking forward to this + more animations (where they are missing plenty right now).
Keep up the good work!

**Comment by Ericky14** [comment]
I have a forked version and added support for blur effect, just sharing it here in case maintainers are interested in making use of that code in here, I don't have enough time right now to contribute here with full support. There could be some more optimizations as well, currently just using full output texture for blur with a multi layer pass.
Still WIP but working

https://github.com/playtron-os/cosmic-comp/commit/c5831351733b76da145b01b8786f97e4c8df4a70

<img width="843" height="359" alt="Image" src="https://github.com/user-attachments/assets/7c281b87-bbaf-4aa4-bfa4-52c4c8cdd4ce" />

**Comment by sethCodesSometimes** [comment]
> > The refraction/liquid glass effect in the kwin-effects-glass thing you linked looked pretty cool. Could such a thing be implemented in cosmic? It seems a little more difficult than just a blur.
> 
> I don't see why not, though I'd imagine it to be low priority until bugs are fixed. Of course, you also gotta make it known to the devs it's something the users want.
> 
> <img width="1366" height="768" alt="Image" src="https://github.com/user-attachments/assets/2da6e727-99c4-4bb3-a312-14654d2d8d8a" />
> The kwin-effects-glass thing is really cool and I hope cosmic implements something to allow this.

This is from the other thread, just putting it here because it's relevant

**Comment by Ryuuusuke** [comment]
this will be cool

**Comment by nalladev** [comment]
Came here after seeing these : 
https://x.com/carlrichell/status/2015831067639636412,
https://x.com/carlrichell/status/2014754933170467129

**Comment by cpuccino** [comment]
Will this also support non-cosmic apps like in hyprland?

**Comment by saberr26** [comment]
> I have a forked version and added support for blur effect, just sharing it here in case maintainers are interested in making use of that code in here, I don't have enough time right now to contribute here with full support. There could be some more optimizations as well, currently just using full output texture for blur with a multi layer pass. Still WIP but working
> 
> [playtron-os@c583135](https://github.com/playtron-os/cosmic-comp/commit/c5831351733b76da145b01b8786f97e4c8df4a70)
> 
> <img alt="Image" width="843" height="359" src="https://private-user-images.githubusercontent.com/21248862/535326957-7c281b87-bbaf-4aa4-bfa4-52c4c8cdd4ce.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzA0MTk5OTUsIm5iZiI6MTc3MDQxOTY5NSwicGF0aCI6Ii8yMTI0ODg2Mi81MzUzMjY5NTctN2MyODFiODctYmJhZi00YWE0LWJmYTQtNTJjNGM4Y2RkNGNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjA2VDIzMTQ1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg0OTdhYmVhYWNkMzNlN2YzN2U1OGE1Zjc4MjA1ZDk5ZTQwZDgyYTkwM2ZjOTc3MzAwM2U1YzE0MTMwZTE0ODUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.oO5uPReoXwqK6hh2_KXe7pGBbRNFAZARdkZMEza6ItE">

how to activate it??

**Comment by Ericky14** [comment]
> > I have a forked version and added support for blur effect, just sharing it here in case maintainers are interested in making use of that code in here, I don't have enough time right now to contribute here with full support. There could be some more optimizations as well, currently just using full output texture for blur with a multi layer pass. Still WIP but working
> > [playtron-os@c583135](https://github.com/playtron-os/cosmic-comp/commit/c5831351733b76da145b01b8786f97e4c8df4a70)
> > <img alt="Image" width="843" height="359" src="https://private-user-images.githubusercontent.com/21248862/535326957-7c281b87-bbaf-4aa4-bfa4-52c4c8cdd4ce.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzA0MTk5OTUsIm5iZiI6MTc3MDQxOTY5NSwicGF0aCI6Ii8yMTI0ODg2Mi81MzUzMjY5NTctN2MyODFiODctYmJhZi00YWE0LWJmYTQtNTJjNGM4Y2RkNGNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjA2VDIzMTQ1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg0OTdhYmVhYWNkMzNlN2YzN2U1OGE1Zjc4MjA1ZDk5ZTQwZDgyYTkwM2ZjOTc3MzAwM2U1YzE0MTMwZTE0ODUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.oO5uPReoXwqK6hh2_KXe7pGBbRNFAZARdkZMEza6ItE">
> 
> how to activate it??

- Clone the repo
- build: `cargo build --release`
- install: `sudo make install`
- reboot or just restart cosmic-comp with `pkill -f cosmic-comp`
- have you apps implement the KDE blur effect and it should work

Config is here if you want to change the blur/backdrop parameters: https://github.com/playtron-os/cosmic-comp/blob/b5f10b186178fab058db7d464a57215550fa5ee6/src/backend/render/blur.rs#L68

**Comment by usbeehu** [comment]
I think prior to building it this should be added too to select the git branch that has the blur thing:
`git checkout 'frosted-glass_noble'`

**Comment by saberr26** [comment]
> I think prior to building it this should be added too to select the git branch that has the blur thing: `git checkout 'frosted-glass_noble'`

Yeah thanks, the blur wasn't working no matter what I did, turns out I was using the wrong Branch

**Comment by nodegin** [comment]
Is there any ETA for this? even like a beta flag feature. Hoping this merged into main ASAP..

**Comment by BrielOtero** [comment]
> Is there any ETA for this? even like a beta flag feature. Hoping this merged into main ASAP..

I am not completely sure, but maybe this is the branch. https://github.com/pop-os/cosmic-comp/tree/frosted-glass_noble

**Comment by adil192** [comment]
Btw if anyone wants to test out frosted glass before it's released, run this mini-script I made to build from source: [install_cosmic_blur.sh](https://github.com/adil192/adil192-linux/blob/d6fee9418726ccbf29b113601d9ef460429ac55d/scripts/install_cosmic_blur.sh).
Update: If you're using Pop!_OS, the easier option is to add the staging repos to apt (see https://github.com/pop-os/cosmic-comp/issues/511#issuecomment-4391467704).

Always check scripts from the internet for safety before running them. Ask AI if you must.

Of course, this is using unreleased code so don't expect any support if you encounter bugs.

**Comment by sethCodesSometimes** [comment]
Thanks! How do you revert the script, though?

**Comment by adil192** [comment]
> Thanks! How do you revert the script, though?

Reinstall the stable packages using your package manager, e.g.:
`sudo apt install --reinstall cosmic-app-library cosmic-applets cosmic-comp cosmic-edit cosmic-files cosmic-greeter cosmic-launcher cosmic-monitor cosmic-notifications cosmic-osd cosmic-panel cosmic-player cosmic-settings cosmic-settings-daemon cosmic-store cosmic-term cosmic-workspaces xdg-desktop-portal-cosmic`

**Comment by mmstick** [comment]
@adil192 You are cloning the git repositories and building them from source. That is not necessary if you add the staging branches to the system.

    sudo apt-manage add popdev:theme-v2
    sudo apt-manage add popdev:frosted-glass

Increase the Pin-Priority of these in /etc/apt/preferences.d/ to 1003 to override packages from `popdev:master` and other staging branches.

**Comment by linux-universe** [comment]
any way to test on arch? Probably by cloning https://aur.archlinux.org/packages/cosmic-comp-git and changing the cloned branch with source= in the file right before building it, correct?

edit: `yay -S cosmic-comp-git --editmenu` and adding #branch=frosted-glass_noble at the end of the git link in source= seems to be working (it is still building)

yeah no, I forgot the other packages needing theme-v2, would work the same way but would also be annoying

**Comment by adil192** [comment]
> [@adil192](https://github.com/adil192) You are cloning the git repositories and building them from source. That is not necessary if you add the staging branches to the system.

Ooh that's really cool. I've updated my comment to mention this for Pop users.
Since I'm on Fedora, building from source is the best option.
@linux-universe You can use my script on arch too

**Comment by Aquabbas** [comment]
How will this be configured/toggled in the final release?

**Comment by adil192** [comment]
> How will this be configured/toggled in the final release?

<img width="937" height="691" alt="Frosted glass settings" src="https://github.com/user-attachments/assets/bf5d6210-3b8c-451c-9bbb-b7d3b1dc1530" />
This is the current implementation of the settings page. I assume it will be similar in the final release.

**Comment by cpuccino** [comment]
Will this work on gtk and kde apps as well? Not expecting it to be on a per component basis but something like hyprland.

**Comment by ShayBox** [comment]
I don't know about GTK, but KDE would likely require setting up Kvantum like every other non-KDE desktop.

**Comment by cpuccino** [comment]
Pretty sure cosmic has/is building their own qt platform.

**Comment by saberr26** [comment]
> Will this work on gtk and kde apps as well? Not expecting it to be on a per component basis but something like hyprland.

Yes, for gtk the gtk theme already follows both opacity and accent from cosmic desktop, for qt (kde apps), it only follows the accent color as of now through libcutecosmic

**Comment by PoisonFrog** [comment]
> > Will this work on gtk and kde apps as well? Not expecting it to be on a per component basis but something like hyprland.
> 
> Yes, for gtk the gtk theme already follows both opacity and accent from cosmic desktop, for qt (kde apps), it only follows the accent color as of now through libcutecosmic

Applications like Kitty Terminal that do not use Qt can be configured to use blur in Plasma using their native configuration files. Would it be possible for an applications to interface with blur in COSMIC like they can in Plasma? Will there be APIs for that or whatever Plasma is doing so developers can make their non-Qt applications be able to use blur in Plasma?

**Comment by adil192** [comment]
> Will this work on gtk and kde apps as well?

Yes, it's using the Wayland [`ext-background-effect-v1`](https://www.phoronix.com/news/Wayland-Background-Effect) protocol.

It's coming in the upcoming Plasma 6.7, COSMIC Epoch 2, Niri 26.04, and more so it will be widely supported across apps/desktops/toolkits.
But there's [no support in GNOME/mutter just yet](https://gitlab.gnome.org/GNOME/mutter/-/work_items/3023), so maybe gnome apps will lag behind. (Don't pester them, they're welcoming contributions for it. Also System76 employees are [helping](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/449) which is very nice to see)

> Applications like Kitty Terminal that do not use Qt can be configured to use blur in Plasma using their native configuration files

Kitty v0.46.2 actually already added support: https://github.com/kovidgoyal/kitty/issues/9534


***

Update: GNOME 51 and GTK 4.24 have added support now :)

**Comment by usbeehu** [comment]
Ghostty 1.4.0 will also add `ext-background-effect-v1` support so that will be a good example on how GTK apps can implement this feature even if Gnome will not support it.

[ghostty-org/ghostty#10721](https://github.com/ghostty-org/ghostty/issues/10721)

**Comment by demonesschurch** [comment]
jaimerai bien que cosmic resemblerai mieux a  Hyprland mes a la sourire non au clavier

**Comment by kleenleen** [comment]
> jaimerai bien que cosmic resemblerai mieux a Hyprland mes a la sourire non au clavier

First of all,  why do you write french here ?
Also, did you even go to school ? There are like 12 fundamental grammar issues in your sentence...

**Comment by leviport** [comment]
Please keep comments on topic. People are following this thread and get notifications for replies here.

**Comment by MythicHH** [comment]
Could there be a way to configure what gets blurred with more granularity? I would like to only have blur for my app launcher (vicinae) and my panels but you currently have to enable it for all windows in order to blur vicinae.

**Comment by defaultUser822** [comment]
I agree with @MythicHH , there should be more granular controls in the config files. (Like which components of COSMIC to enable the blur for)

**Comment by imnotliveyet** [comment]
The moment this is implemented, I'm installing Pop!_OS as a permanent fixture in my fleet of scrap PCs

**Comment by MythicHH** [comment]
> The moment this is implemented, I'm installing Pop!_OS as a permanent fixture in my fleet of scrap PCs

you can already use it if you enable the correct repositories. there is a guide on one of the reddit pages somewhere.

**Comment by RealCrystalNight** [comment]
> > The moment this is implemented, I'm installing Pop!_OS as a permanent fixture in my fleet of scrap PCs
> 
> you can already use it if you enable the correct repositories. there is a guide on one of the reddit pages somewhere.

He means in a stable release, this is triaged for epoch 2. Not sure when that releases though.

**Comment by Shruichan** [comment]
Is this still being worked on? Frosted glass / blur is basically the one thing I miss coming from KDE. Happy to help test or pick up part of it if that'd actually be useful.

**Comment by nalladev** [comment]
Just curious—since COSMIC already supports importing theme files that change colors across the desktop, would it be possible for visual styles (like Apple's new Liquid Glass design) to be distributed as importable themes as well? Or would effects like blur and translucency always require manual configuration instead?

**Comment by usbeehu** [comment]
@Shruichan It is in fact pretty close to release.

**Comment by usbeehu** [comment]
@nalladev Transparency (alpha channel) values are in the theme file, so I think it can be enabled or disabled through that but I'm not entirely sure.

But your question raised me a lot of ideas. There is Cosmic Tweaks that aims to do more advanced stuffs too. It has a feature that lets you switch between predefined layouts. Now it would be super cool if that could mimic other desktops the most accurate way by storing dock/panel position, icon themes, color themes, frosted glass, etc. basically in a 1 click solution, similar to ZorinOS for example. Now that would require more and better applets but if things go well, it would be possible to perfectly recreate a Windows 7, macOS, Unity or KDE desktop. But there is no one click solution for that.

**Comment by BryanMaker** [comment]
> [@nalladev](https://github.com/nalladev) Transparency (alpha channel) values are in the theme file, so I think it can be enabled or disabled through that but I'm not entirely sure.
> 
> But your question raised me a lot of ideas. There is Cosmic Tweaks that aims to do more advanced stuffs too. It has a feature that lets you switch between predefined layouts. Now it would be super cool if that could mimic other desktops the most accurate way by storing dock/panel position, icon themes, color themes, frosted glass, etc. basically in a 1 click solution, similar to ZorinOS for example. Now that would require more and better applets but if things go well, it would be possible to perfectly recreate a Windows 7, macOS, Unity or KDE desktop. But there is no one click solution for that.

Please don't let it end up as complicated as KDE 😂️

**Comment by usbeehu** [comment]
@BryanMaker How is it complicated? The point would be that a single click could add/remove applets, modify their placement, could enable live wallpaper, change icon theme, etc. without having you to modify all of these manually.

**Comment by leviport** [comment]
Let's keep this thread on topic, please. People are getting notifications for posts here.
